lib/system.mli

```diff
@@ -87,14 +88,35 @@ val with_magic_number_check : ('a -> 'b) -> 'a -> 'b
 val marshal_out : out_channel -> 'a -> unit
 val marshal_in : string -> in_channel -> 'a
 
-(** Clones of Digest.output and Digest.input (with nice error message) *)
+module ObjFile :
+sig
 
-val digest_out : out_channel -> Digest.t -> unit
-val digest_in : string -> in_channel -> Digest.t
+val magic_number : int32
 
-val marshal_out_segment : string -> out_channel -> 'a -> unit
-val marshal_in_segment : string -> in_channel -> 'a * int * Digest.t
-val skip_in_segment : string -> in_channel -> int * Digest.t
+type segment = {
+  name : string;
+  pos : int64;
+  len : int64;
+  hash : Digest.t;
+}
+
+type in_handle
+type out_handle
+
+val open_in : file:string -> in_handle
+val close_in : in_handle -> unit
+val marshal_in_segment : in_handle -> segment:string -> 'a * Digest.t
+val get_segment : in_handle -> segment:string -> segment
+val segments : in_handle -> segment CString.Map.t
+
+val open_out : file:string -> out_handle
+val close_out : out_handle -> unit
+val marshal_out_segment : out_handle -> segment:string -> 'a -> unit
+val marshal_out_binary : out_handle -> segment:string -> out_channel * (unit -> unit)
```

Comment:
can one interleave these two APIs and the closure? I guess no

Reply:
Nope, the comment says that you can ony "write to it", which means the low-level write functions of OCaml.

Reply:
I think you should make a runtime check that marshal_out_segment is not called while the "session" for the binary segment is open.

The comment does not say that, I read it as "you should not seek or close".

Reply:
I am not too fond of dynamic tests when this is a clear API restriction. FTR, this function is not used currently but I need it for further changes regarding the storage of opaque proofs.

Reply:
The comment is not clear. Let me propose:
```
[marshal_out_binary oh segment] is a low level, stateful, API returning [oc, stop]. Once called no other 
API can be used on the same [oh] and only [Stdlib.output_*] APIs should be used on [oc]. [stop ()] must be invoked
in order to signal that all data was written to [oc] (which should not be used afterwards). Only after calling [stop] the other API can be used on [oh].
```